### ISSUE TITLE
Promises

### ISSUE THREAD

**Issue by LukeTim** [fresc81/node-winreg#38]

**Comment by fresc81** [comment]
Promises are not yet supported. But I like the idea. In the meantime maybe https://github.com/caolan/async would come in handy. I usually use it when dealing with asynchronous code.

**Comment by RX14** [comment]
There's always bluebird's promisify. I found it hard to use promises at all without bluebird, but that was over a year ago...

**Comment by fresc81** [comment]
Yeah, looks like you could even use bluebird to promisify other Promise unaware APIs like winreg's for example. This probhably suits your usecase even more, @LukeTim .
http://bluebirdjs.com/docs/features.html#promisification-on-steroids

**Comment by fresc81** [comment]
Would be great if you can give some feedback if bluebirdjs works in combination with winreg.

**Comment by LukeTim** [comment]
So it's been a while but I've finally got some down time to take another look at this app. I tried using bluebird but it doesn't seem to work. I just get told that "keysAsync" doesn't exist.

I might give async a go, but I am not sure it will help. I need a way to have a piece of code that is guaranteed to run *after* all of my calls to `.get()` are completed, so that I can build an object and push it on to an array.

**Comment by fresc81** [comment]
Thanks for your feedback. Would aliases for the winreg functions help in this case? e.g. <code>Registry.prototype.getAsync = Registry.prototype.get</code>
You should give async definitely a try. Performing async tasks and then call callbacks is basically what it's useful for.

**Comment by dkudrin** [comment]
@fresc81 
It works only this way:
`
    import { promisifyAll } from 'bluebird'

    let regKey = new Registry({
          hive: Registry.HKCU,
          key:  '\\Software\\Microsoft\\Windows\\CurrentVersion\\Run'
        })

    promisifyAll(regKey)

    let vals = await regKey.valuesAsync()
    console.log(vals)

`

**Comment by tgds** [comment]
@fresc81, first of all, thank you for the library, works great!
I prefer working with Promises as well, so I took the liberty, [forked your project](https://github.com/tgds/node-winreg) and added support for native Promises (node >= 0.12.18). Each method now returns a Promise if no callback is specified.

Let me know if you'd be interested in merging these changes. I'd send you a pull request, but it seems messy as there are some issues with whitespace and the diffs are large, so I'm asking here first before I start cleaning it up. Perhaps, there's still some work needed on the documentation part as well.

**Comment by LukeTim** [comment]
@dkudrin Thank you so much for that. I managed to get promises to work with bluebird now. An absolute godsend. This project had been sitting on the back burner for months and now I'm making a lot of progress.

**Comment by neriyaco** [comment]
I know it's old but I just released a promise-based package for this project [winreg-promise](https://github.com/xpodev/winreg-promise)
